File: backend/app/schemas/volume_intelligence.py
"""DTOs mirroring analytics/volume_intelligence/models.py's dataclasses.
Served only by GET /api/v1/volume-intelligence/{symbol} -- not part of the
hot-polled /dashboard/{symbol}/latest response, since this needs a
multi-day (~60 trading day) candle fetch and full recomputation of 14
metrics, much heavier than the regular poll.
"""

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel

BaselineGroupLiteral = Literal["yesterday", "last_5_days", "last_20_days", "same_weekday", "expiry_day", "monthly_expiry_day"]
RvolLabelLiteral = Literal["Above Average", "Average", "Below Average"]
DominantSideLiteral = Literal["buy", "sell", "balanced"]
AccelerationLabelLiteral = Literal["Accelerating", "Stable", "Decelerating"]
MomentumLabelLiteral = Literal["Strong Buy Momentum", "Buy Momentum", "Neutral", "Sell Momentum", "Strong Sell Momentum"]
InstitutionalLabelLiteral = Literal["Minimal", "Low", "Moderate", "High", "Very High"]
VolumeTrendLabelLiteral = Literal["Strong Increasing", "Increasing", "Stable", "Decreasing", "Strong Decreasing"]
VolumeCharacterLabelLiteral = Literal["Accumulation", "Distribution", "Markup", "Markdown", "Climactic", "Quiet-Consolidation"]
ExhaustionDirectionLiteral = Literal["up", "down"]
ResemblanceLabelLiteral = Literal[
    "accumulation-like sessions",
    "distribution-like sessions",
    "climactic/high-volume sessions",
    "quiet/low-volume sessions",
    "mixed/typical sessions",
]
ForecastConfidenceLiteral = Literal["Low", "Medium", "High"]
DailyComparisonLabelLiteral = Literal["Much Higher", "Higher", "Similar", "Lower", "Much Lower"]
PriceDirectionLiteral = Literal["up", "down", "flat"]


class RvolBaselineResultDTO(BaseModel):
    group: BaselineGroupLiteral
    interval_rvol_pct: float | None
    cumulative_rvol_pct: float | None
    label: RvolLabelLiteral | None
    sample_days: int


class RvolReadingDTO(BaseModel):
    by_baseline: dict[str, RvolBaselineResultDTO]
    primary: RvolBaselineResultDTO | None


class VolumeAccelerationDTO(BaseModel):
    recent_avg_volume: float
    prior_avg_volume: float
    ratio: float | None
    pct_change: float | None
    label: AccelerationLabelLiteral


class VolumeSpikeDTO(BaseModel):
    is_spike: bool
    multiple: float | None
    baseline_source: Literal["historical_20d", "intraday_rolling"] | None
    baseline_volume: float | None


class VolumeDryUpDTO(BaseModel):
    is_dryup: bool
    fraction: float | None
    baseline_source: Literal["historical_20d", "intraday_rolling"] | None
    baseline_volume: float | None


class BuySellDominanceDTO(BaseModel):
    window_minutes: int
    buy_volume: float
    sell_volume: float
    dominance_ratio: float
    dominant_side: DominantSideLiteral
    consecutive_dominant_minutes: int


class CumulativePressureDTO(BaseModel):
    cum_buy_volume: float
    cum_sell_volume: float
    net_pressure: float
    pressure_ratio: float


class VolumeMomentumDTO(BaseModel):
    ema_signed_volume: float
    normalized_score: float
    streak_minutes: int
    label: MomentumLabelLiteral


class InstitutionalParticipationDTO(BaseModel):
    score: int
    label: InstitutionalLabelLiteral
    rvol_component: float
    blockiness_component: float
    dominance_component: float


class AbsorptionSignalDTO(BaseModel):
    detected: bool
    range_ratio: float | None
    volume_multiple: float | None
    side_hint: Literal["buy_absorption", "sell_absorption", "undetermined"]


class ExhaustionSignalDTO(BaseModel):
    detected: bool
    direction: ExhaustionDirectionLiteral | None
    move_over_window: float | None
    volume_multiple: float | None
    wick_ratio: float | None


class VolumeTrendDTO(BaseModel):
    window_minutes: int
    pct_change: float | None
    label: VolumeTrendLabelLiteral


class VolumeCharacterDTO(BaseModel):
    label: VolumeCharacterLabelLiteral
    rationale: str


class SimilarDayDTO(BaseModel):
    session_date: date
    distance: float
    similarity: float
    dominant_side: DominantSideLiteral
    total_volume_ratio: float


class HistoricalSimilarityDTO(BaseModel):
    top_days: list[SimilarDayDTO]
    resemblance_label: ResemblanceLabelLiteral | None
    n_days_compared: int


class NextIntervalForecastDTO(BaseModel):
    horizon_minutes: int
    probability_continuation: float
    probability_reversal: float
    confidence: ForecastConfidenceLiteral
    supporting_factors: list[str]
    composite_score: float


class VolumeNarrativeDTO(BaseModel):
    headline: str
    observations: list[str]


class DailyVolumeComparisonDTO(BaseModel):
    session_date: date
    volume_as_of: float
    prior_day_volume_as_of: float | None
    pct_change: float | None
    label: DailyComparisonLabelLiteral | None
    interpretation: str


class DailyVolumeTrendDTO(BaseModel):
    elapsed_minutes: int
    days: list[DailyVolumeComparisonDTO]


class SignificantIntervalDTO(BaseModel):
    start_time: datetime
    end_time: datetime
    interval_volume: float
    baseline_volume: float | None
    pct_change: float | None
    multiple: float | None
    dominant_side: DominantSideLiteral
    price_direction: PriceDirectionLiteral
    institutional_note: str
    trend_note: str


class VolumeIntelligenceDTO(BaseModel):
    symbol: str
    as_of: datetime | None
    rvol: RvolReadingDTO | None
    acceleration: VolumeAccelerationDTO | None
    dominance: BuySellDominanceDTO | None
    cumulative_pressure: CumulativePressureDTO | None
    momentum: VolumeMomentumDTO | None
    institutional: InstitutionalParticipationDTO | None
    spike: VolumeSpikeDTO | None
    dryup: VolumeDryUpDTO | None
    absorption: AbsorptionSignalDTO | None
    exhaustion: ExhaustionSignalDTO | None
    trend: VolumeTrendDTO | None
    character: VolumeCharacterDTO | None
    similarity: HistoricalSimilarityDTO | None
    forecast: NextIntervalForecastDTO | None
    narrative: VolumeNarrativeDTO | None
    daily_volume_trend: DailyVolumeTrendDTO | None
    significant_intervals: list[SignificantIntervalDTO]

File: frontend/src/types/volumeIntelligence.ts
// Mirrors backend/app/schemas/volume_intelligence.py exactly.

export type BaselineGroup = "yesterday" | "last_5_days" | "last_20_days" | "same_weekday" | "expiry_day" | "monthly_expiry_day";
export type RvolLabel = "Above Average" | "Average" | "Below Average";
export type DominantSide = "buy" | "sell" | "balanced";
export type AccelerationLabel = "Accelerating" | "Stable" | "Decelerating";
export type MomentumLabel = "Strong Buy Momentum" | "Buy Momentum" | "Neutral" | "Sell Momentum" | "Strong Sell Momentum";
export type InstitutionalLabel = "Minimal" | "Low" | "Moderate" | "High" | "Very High";
export type VolumeTrendLabel = "Strong Increasing" | "Increasing" | "Stable" | "Decreasing" | "Strong Decreasing";
export type VolumeCharacterLabel = "Accumulation" | "Distribution" | "Markup" | "Markdown" | "Climactic" | "Quiet-Consolidation";
export type ExhaustionDirection = "up" | "down";
export type ResemblanceLabel =
  | "accumulation-like sessions"
  | "distribution-like sessions"
  | "climactic/high-volume sessions"
  | "quiet/low-volume sessions"
  | "mixed/typical sessions";
export type ForecastConfidence = "Low" | "Medium" | "High";

export interface RvolBaselineResultDTO {
  group: BaselineGroup;
  interval_rvol_pct: number | null;
  cumulative_rvol_pct: number | null;
  label: RvolLabel | null;
  sample_days: number;
}

export interface RvolReadingDTO {
  by_baseline: Partial<Record<BaselineGroup, RvolBaselineResultDTO>>;
  primary: RvolBaselineResultDTO | null;
}

export interface VolumeAccelerationDTO {
  recent_avg_volume: number;
  prior_avg_volume: number;
  ratio: number | null;
  pct_change: number | null;
  label: AccelerationLabel;
}

export interface VolumeSpikeDTO {
  is_spike: boolean;
  multiple: number | null;
  baseline_source: "historical_20d" | "intraday_rolling" | null;
  baseline_volume: number | null;
}

export interface VolumeDryUpDTO {
  is_dryup: boolean;
  fraction: number | null;
  baseline_source: "historical_20d" | "intraday_rolling" | null;
  baseline_volume: number | null;
}

export interface BuySellDominanceDTO {
  window_minutes: number;
  buy_volume: number;
  sell_volume: number;
  dominance_ratio: number;
  dominant_side: DominantSide;
  consecutive_dominant_minutes: number;
}

export interface CumulativePressureDTO {
  cum_buy_volume: number;
  cum_sell_volume: number;
  net_pressure: number;
  pressure_ratio: number;
}

export interface VolumeMomentumDTO {
  ema_signed_volume: number;
  normalized_score: number;
  streak_minutes: number;
  label: MomentumLabel;
}

export interface InstitutionalParticipationDTO {
  score: number;
  label: InstitutionalLabel;
  rvol_component: number;
  blockiness_component: number;
  dominance_component: number;
}

export interface AbsorptionSignalDTO {
  detected: boolean;
  range_ratio: number | null;
  volume_multiple: number | null;
  side_hint: "buy_absorption" | "sell_absorption" | "undetermined";
}

export interface ExhaustionSignalDTO {
  detected: boolean;
  direction: ExhaustionDirection | null;
  move_over_window: number | null;
  volume_multiple: number | null;
  wick_ratio: number | null;
}

export interface VolumeTrendDTO {
  window_minutes: number;
  pct_change: number | null;
  label: VolumeTrendLabel;
}

export interface VolumeCharacterDTO {
  label: VolumeCharacterLabel;
  rationale: string;
}

export interface SimilarDayDTO {
  session_date: string;
  distance: number;
  similarity: number;
  dominant_side: DominantSide;
  total_volume_ratio: number;
}

export interface HistoricalSimilarityDTO {
  top_days: SimilarDayDTO[];
  resemblance_label: ResemblanceLabel | null;
  n_days_compared: number;
}

export interface NextIntervalForecastDTO {
  horizon_minutes: number;
  probability_continuation: number;
  probability_reversal: number;
  confidence: ForecastConfidence;
  supporting_factors: string[];
  composite_score: number;
}

export interface VolumeNarrativeDTO {
  headline: string;
  observations: string[];
}

export interface VolumeIntelligenceDTO {
  symbol: string;
  as_of: string | null;
  rvol: RvolReadingDTO | null;
  acceleration: VolumeAccelerationDTO | null;
  dominance: BuySellDominanceDTO | null;
  cumulative_pressure: CumulativePressureDTO | null;
  momentum: VolumeMomentumDTO | null;
  institutional: InstitutionalParticipationDTO | null;
  spike: VolumeSpikeDTO | null;
  dryup: VolumeDryUpDTO | null;
  absorption: AbsorptionSignalDTO | null;
  exhaustion: ExhaustionSignalDTO | null;
  trend: VolumeTrendDTO | null;
  character: VolumeCharacterDTO | null;
  similarity: HistoricalSimilarityDTO | null;
  forecast: NextIntervalForecastDTO | null;
  narrative: VolumeNarrativeDTO | null;
}
